CMakeLists.txt

```diff
@@ -323,3 +324,48 @@ if(VLLM_GPU_LANG STREQUAL "CUDA" OR VLLM_GPU_LANG STREQUAL "HIP")
   add_dependencies(default _moe_C)
 
 endif()
+
+#
+# Build flash attention from source
+#
+# IMPORTANT: This has to be the last thing we do, because flash-attention uses the same macros/fuctions as vLLM.
+# Because functions all belong to the global scope, flash-attention's funcitons overwrite vLLMs.
+# They should be identical but if they aren't, this is a massive footgun.
+#
+
+# If VLLM_FLASH_ATTN_SRC_DIR is set, vllm-flash-attn is installed from that directory instead of downloading.
+# This is to enable local development of vllm-flash-attn within vLLM.
+# It can be set as an environment variable or passed as a cmake argument.
+if (DEFINED ENV{VLLM_FLASH_ATTN_SRC_DIR})
+  set(VLLM_FLASH_ATTN_SRC_DIR $ENV{VLLM_FLASH_ATTN_SRC_DIR})
+endif()
+
+if(VLLM_FLASH_ATTN_SRC_DIR)
+  FetchContent_Declare(flash-attention SOURCE_DIR ${VLLM_FLASH_ATTN_SRC_DIR})
+else()
+  # TODO(luka) use vllm-project version
+  FetchContent_Declare(
+          flash-attention
+          GIT_REPOSITORY https://github.com/neuralmagic/flash-attention.git
```

Comment:
fix this? and do we use the `main` branch or pin the commit?

Reply:
I think we pin the commit. This can only be fixed once the vllm-flash-attn PR #19 is merged

Reply:
the link is already merged?

Reply:
Sorry I meant [this PR](https://github.com/vllm-project/flash-attention/pull/19) on the vllm-flash-attn repo